gui-js/apps/minsky-electron/src/app/managers/CommandsManager.ts

```diff
@@ -1186,14 +1186,60 @@ export class CommandsManager {
       }
     });
   }
-  
-  static upgrade() {
-    let window=WindowManager.createWindow({
+
+  static startCSVDownload(payload: DownloadCSVPayload) {
+    const window=this.createDownloadWindow();
+
+    const splitUrl = payload.url.split('.');
+    const extension = splitUrl.length === 1 ? 'csv' : splitUrl.pop();
+    const savePath = join(tmpdir(),`downloaded.${extension}`);
+    window.webContents.session.on('will-download', (e,i,w) => this.downloadCSV(e,i,w,savePath));
+    window.webContents.downloadURL(payload.url);
+    return savePath;
```

Comment:
This is returning too early. electron.events is awaiting on this, will return the file name to electronservice, which then returns to importCSV, which will then attempt to open the file, which is still in the process of being downloaded.

So we need to make this method async, create a promise here to be returned, and arrange for the promise to be resolved/rejected in the done handler of webContents.

Granted, Promises are not exactly the easiest APIs to wrap you head around, but there's an example of doing exactly this in CommandsManager.newPubTab().